File: OsmAnd-java/src/main/java/net/osmand/data/LatLon.java
package net.osmand.data;

import net.osmand.util.MapUtils;

import java.io.Serializable;

public class LatLon implements Serializable {

	private static final long serialVersionUID = 1811582709897737392L;
	private final double latitude;
	private final double longitude;

	public LatLon(double latitude, double longitude) {
		this.latitude = latitude;
		this.longitude = longitude;
	}

	public double getLatitude() {
		return latitude;
	}

	public double getLongitude() {
		return longitude;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		int temp;
		temp = (int) Math.floor(latitude * 10000);
		result = prime * result + temp;
		temp = (int) Math.floor(longitude * 10000);
		result = prime * result + temp;
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;

		LatLon other = (LatLon) obj;
		return MapUtils.areLatLonEqual(this, other);
	}

	@Override
	public String toString() {
		return "Lat " + ((float) latitude) + " Lon " + ((float) longitude);
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/data/LatLon.kt
package net.osmand.shared.data

import kotlinx.serialization.Serializable
import net.osmand.shared.util.MapUtils
import kotlin.math.floor

@Serializable
data class LatLon(val latitude: Double, val longitude: Double) {

	override fun hashCode(): Int {
		val prime = 31
		var result = 1
		var temp: Int = floor(latitude * 10000).toInt()
		result = prime * result + temp
		temp = floor(longitude * 10000).toInt()
		result = prime * result + temp
		return result
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || this::class != other::class) return false
		other as LatLon
		return MapUtils.areLatLonEqual(this, other)
	}

	override fun toString(): String {
		return "Lat ${latitude.toFloat()} Lon ${longitude.toFloat()}"
	}
}
